### PR TITLE
FEATURE: User management for non-admins

### PR DESCRIPTION
**Pull request by dimaip**
This introduces a new role `Neos.Neos:UserManager`. When an editor has that role, he is able to manage users with roles not exceeding his own, without being an Administrator.

peek 2018-12-15 23-40

### CODE REVIEW

**Comment by dimaip** [comment]
Since I got some positive feedback of this feature, I tried to finalize it. Please provide reviews!

File: Neos.Neos/Resources/Private/Translations/en/Modules.xlf [hunk 1]
```diff
@@ -428,6 +428,12 @@
 			<trans-unit id="users.tooltip.edit" xml:space="preserve">
 				<source>Edit user</source>
 			</trans-unit>
+			<trans-unit id="users.tooltip.editDisallowed" xml:space="preserve">
```
Comment (daniellienert) [inline Neos.Neos/Resources/Private/Translations/en/Modules.xlf]: If I don't get it wrong, the entries for alle the new flash messages are missing.

Comment (dimaip) [reply Neos.Neos/Resources/Private/Translations/en/Modules.xlf]: Sorry, I don't get what you mean. I added them for English, must I add them for every language?

File: Neos.Neos/Classes/Controller/Module/Administration/UsersController.php [hunk 2]
```diff
@@ -79,6 +84,8 @@ protected function initializeAction()
             }
         }
         $this->currentUser = $this->userService->getCurrentUser();
+        $adminRole = $this->policyService->getRole('Neos.Neos:Administrator');
```
[the remark this answers is not in the document]
Comment (dimaip) [reply Neos.Neos/Classes/Controller/Module/Administration/UsersController.php]: Cool, thanks for the tip

File: Neos.Neos/Classes/Controller/Module/Administration/UsersController.php [hunk 3]
```diff
@@ -88,9 +95,19 @@ protected function initializeAction()
      */
     public function indexAction()
     {
+        $users = $this->userService->getUsers();
+        $usersWithMeta = [];
+        foreach ($users as $user) {
+            $userRoles = $user->getAccounts()[0]->getRoles();
```
[the remark this answers is not in the document]
Comment (dimaip) [reply Neos.Neos/Classes/Controller/Module/Administration/UsersController.php]: Can make it public ofc

**Comment by dimaip** [comment]
Wow @daniellienert, thanks for your thorough code review, tried to adjust to your comments. Please take a look again when you got the time!

File: Neos.Neos/Resources/Private/Translations/en/Modules.xlf (hunk 1 above)
Comment (daniellienert) [reply Neos.Neos/Resources/Private/Translations/en/Modules.xlf]: No, not for every language, but you need to register the sentence with its id for english and use this id in the flashMessage, like it is done here for example:

- https://github.com/neos/neos-development-collection/blob/4d5219277da26e62f6ce5335d74a1ddd1ac1a86d/Neos.Media.Browser/Classes/Controller/TagController.php#L75
- https://github.com/neos/neos-development-collection/blob/7584a1d6ef3666c471a5ae9147dd585c6cf77a7c/Neos.Media.Browser/Resources/Private/Translations/en/Main.xlf#L400

Sadly, it seems that a lot of flash messages in the user module are not translated yet. But I would like to not add more untranslated messages.

**Comment by albe** [comment]
Folks, that looks like a very nice feature - any chance of getting it finished up for 5.2? Two weeks left to have it merge ready and from what I can tell it's already pretty close

**Comment by albe** [comment]
So, not this time. But hopefully for next release :)

File: Neos.Neos/Classes/Controller/Module/Administration/UsersController.php [hunk 4]
```diff
@@ -139,8 +161,14 @@ public function newAction(User $user = null)
      */
     public function createAction($username, array $password, User $user, array $roleIdentifiers, $authenticationProviderName = null)
     {
-        $this->userService->addUser($username, $password[0], $user, $roleIdentifiers, $authenticationProviderName);
-        $this->addFlashMessage('The user "%s" has been created.', 'User created', Message::SEVERITY_OK, [htmlspecialchars($username)], 1416225561);
+        $currentUserRoles = $this->userService->getAllRoles($this->currentUser);
+        $isCreationAllowed = count(array_diff($roleIdentifiers, $currentUserRoles)) === 0;
```
Comment (daniellienert) [inline Neos.Neos/Classes/Controller/Module/Administration/UsersController.php:165]: I guess, that will not work. That only allows creating users, which has exact the same roles then I have.

File: Neos.Neos/Classes/Controller/Module/Administration/UsersController.php [hunk 5]
```diff
@@ -317,4 +375,32 @@ protected function getAuthenticationProviders()
...
+     * @param User $user
+     */
+    protected function isEditingAllowed(User $user): bool
+    {
+        if ($this->isAdministrator) {
+            return true;
+        }
+        $currentUserRoles = $this->userService->getAllRoles($this->currentUser);
+        $userRoles = $this->userService->getAllRoles($user);
+        return count(array_diff($userRoles, $currentUserRoles)) === 0;
```
Comment (daniellienert) [inline Neos.Neos/Classes/Controller/Module/Administration/UsersController.php:404]: That would mean, editing is only allowed when the user to edit has the same roles than me. That makes no sense to me.

**Comment by daniellienert** [comment]
@dimaip, I rebased and refactored the feature. Please have a look at the above comments. It seems to me, that they are logically wrong. IMHO it needs to be changed to "I have alle the roles of the user I want to edit but may have more"

File: Neos.Neos/Classes/Controller/Module/Administration/UsersController.php [hunk 6]
```diff
@@ -277,6 +327,10 @@ public function createElectronicAddressAction(User $user, ElectronicAddress $ele
      */
     public function deleteElectronicAddressAction(User $user, ElectronicAddress $electronicAddress)
     {
+        if (!$this->isEditingAllowed($user)) {
+            $this->addFlashMessage('Not allowed to delete an electronic address for the user "%s".', 'User email deletion denied', Message::SEVERITY_ERROR, [$user->getName()->getFullName()], 1416225567);
```
[the remark this answers is not in the document]
Comment (kdambekalns) [reply Neos.Neos/Classes/Controller/Module/Administration/UsersController.php:344]: No, exception codes designate a unique place, not "the same exception". Has never been different.

File: Neos.Neos/Resources/Private/Translations/en/Modules.xlf (hunk 1 above)
Comment (kdambekalns) [reply Neos.Neos/Resources/Private/Translations/en/Modules.xlf:431]: AFAIS the items added are in the XLIFF file, everything else was already there (and if it was missing in XLIFF already, that is nothing that should block this one…

**Comment by kdambekalns** [comment]
@daniellienert Feel free to check again and merge (ASAP)…

**Comment by daniellienert** [comment]
@daniellienert will have the time today for another round of testing. Will be merged till end of the day.

File: Neos.Neos/Configuration/Policy.yaml [hunk 7]
```diff
@@ -258,6 +258,9 @@ roles:
 
   'Neos.Neos:UserManager':
     privileges:
+      -
+        privilegeTarget: 'Neos.Neos:Backend.GeneralAccess'
```
Comment (kdambekalns) [inline Neos.Neos/Configuration/Policy.yaml:262]: Ah, let me guess: That makes login work, even if *only* `UserManager` as assigned… 👍

Comment (daniellienert) [reply Neos.Neos/Configuration/Policy.yaml:262]: Yeah, with the current implementation of the backend login and redirect you will end up in the content module where you don't have any access. But if we fix that (or if you know the correct url) you can now access the user module

**Comment by kdambekalns** [comment]
👏

**Comment by dimaip** [comment]
Woohoo, thanks so much guys for reviving this PR! ❤️

**Pull request merged**